Add support for EngagingNetworks

bruno opened the pull request:
This is like JacksonRiver and other CRM/PaymentGateway concentrators. It's not technically a gateway, as it uses Vantiv/Worldpay in the background, combined with BBCRM.

The gateway supports creditcard and directdebit `purchase`, however given the comment above, you will see that the implementation contains support for recurring billing for each payment method, however those are handled by EN and not sent/expected differently. I've implemented this to match the approach we took with Convio, i.e. we expose only one method and the CRM does all the decision based on the data we pass.

EN supplies an `api_key` that we pass on to retrieve a session token during the authentication call, we then use that token in the headers of our subsequent interaction.

```Shell
$ docker run -it --rm -v "$PWD":/app waysact/active_merchant ruby -Itest test/unit/gateways/engaging_networks_test.rb
Loaded suite test/unit/gateways/engaging_networks_test
Started
.....
Finished in 0.031073083 seconds.
-----------------------------------------------------------------------------------------------------
5 tests, 19 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
-----------------------------------------------------------------------------------------------------
160.91 tests/s, 611.46 assertions/s
```

```Shell
$ docker run -it --rm -v "$PWD":/app waysact/active_merchant ruby -Itest test/remote/gateways/remote_engaging_networks_test.rb
Loaded suite test/remote/gateways/remote_engaging_networks_test
Started
.........
Finished in 18.350281955 seconds.
-----------------------------------------------------------------------------------------------------
9 tests, 18 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
-----------------------------------------------------------------------------------------------------
```

Refs waysact/evergiving#6789

bruno committed:
Add support for EngagingNetworks

This is like JacksonRiver and other CRM/PaymentGateway concentrators. It's not technically a gateway, as it uses Vantiv/Worldpay in the background, combined with BBCRM.

The gateway supports creditcard and directdebit `purchase`, however given the comment above, you will see that the implementation contains support for recurring billing for each payment method, however those are handled by EN and not sent/expected differently. I've implemented this to match the approach we took with Convio, i.e. we expose only one method and the CRM does all the decision based on the data we pass.

EN supplies an `api_key` that we pass on to retrieve a session token during the authentication call, we then use that token in the headers of our subsequent interaction.

Refs waysact/evergiving#6789

henare commented on lib/active_merchant/billing/gateways/engaging_networks.rb:
```diff
@@ -0,0 +1,346 @@
+            /[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}/i,
+            '[FILTERED]'
+          ).gsub(
+            /(ens-auth-token: )([0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12})/i,
```
I think this would be captured by the previous regex anyway so it might be redundant (still, it's not harmful :slightly_smiling_face:).

bruno replied:
Good point, yes, it would ... I will leave it just to be more explicit...

henare commented on lib/active_merchant/billing/gateways/engaging_networks.rb:
```diff
@@ -0,0 +1,346 @@
+      # if there's no donor email.
+      #
+      # Format: timestamp@fakeemail(custom_label).com
+      def email(email, options)
```
Rather than passing in the whole `options` hash and then extracting a value from it, it might be clearer to have the second parameter be the custom label itself?

henare commented on lib/active_merchant/billing/gateways/engaging_networks.rb:
```diff
@@ -0,0 +1,346 @@
+      #
+      # `questions` is a whole set of metadata, more like a survey
+      # however all the fields are hardcoded on the CRM side, thus
+      # why we use the constant to hold them.
```
Are these questions hardcoded by Engaging Networks or our customer? If they're our customer's then it might be a good idea to push this into configuration rather than hardcoding it into ActiveMerchant.

If it's easier to do this and worry about it when another customer wants to use Engaging Networks then that totally makes sense too :wink:

bruno replied:
That's a very good point. It's definitely a customer requirement and not dictated by EngagingNetworks (it's the CRM in the backend). I will make those changes...

henare commented on lib/active_merchant/billing/gateways/engaging_networks.rb:
```diff
@@ -0,0 +1,346 @@
+        case e.response.code
+        when '401'
+          return Response.new(false, 'Invalid credentials', {}, test: test?)
+        end
```
Do we want to re-raise if it's not a 401? Otherwise I think we're silently rescuing and not doing anything with the exception.

bruno replied:
Good catch!

henare commented on lib/active_merchant/billing/gateways/engaging_networks.rb:
```diff
@@ -0,0 +1,346 @@
+          else
+            "#{response['status']}|#{response['type']}"
+          end
+        end
```
Rather than a `case` and nested `if`, which I find slightly hard-to-read, this could be in the single `case` statement:

```ruby
        case 
        when action == 'auth'
          response['ens-auth-token']
        when !success_from(action, response)
          response['error']
        else
          "#{response['status']}|#{response['type']}"
        end
```

bruno replied:
Oh! yeah, that's nice ...

henare commented on lib/active_merchant/billing/gateways/engaging_networks.rb:
```diff
@@ -0,0 +1,346 @@
+          parameters.to_s
+        else
+          parameters.to_json
+        end
```
A few of these `case` statements could be simpler and more concise as using the ternary operator, e.g.

```ruby
action == 'auth' ? parameters.to_s : parameters.to_json
```

henare reviewed (commented):
Very nice :ok_hand: 

I wasn't able to run the remote tests because of the gateway's IP whitelisting but you've run them and they work so :sunglasses: (It was great to have the console logs that included the commands to see those tests run BTW :+1:).

I've added a few comments, they're mainly nitpicks so feel free to ignore the ones you disagree with :slightly_smiling_face: The only one I would flag :flags: is not re-raising on `ResponseError` when the error's not `401`. That seems important to fix.

bruno committed:
Refactor email method

Pass the label already to the method, so we don't have to pass around a large object anyway.

Refs waysact/evergiving#6789

bruno committed:
Raise in case it's not a 401 response

Good catch @henare ... re-raise in case it's not the expected ResponseError for a failed login.

Refs waysact/evergiving#6789

bruno committed:
Refactor case statements

Based on the review feedback, making the case statements more readable by, in most cases, turning them into ternary operators or removing nested conditionals.

Refs waysact/evergiving#6789

bruno committed:
Remove extra `]` that was leftover during refactor

That'll teach me to run the test suite before committing and pushing ...

Refs waysact/evergiving#6789

bruno committed:
Refactor `questions` options

Extracted them into a set of options that we expect the caller to know (in this case, it'll be built into the application based on the configuration for the given customer, as each customer might have a different set of questions/metadata).

Refs waysact/evergiving#6789

bruno committed:
Prettier-ify

Refs waysact/evergiving#6789

bruno commented:
Re-ran the tests:

```
❯ docker run -it --rm -v "$PWD":/app waysact/active_merchant ruby -Itest test/remote/gateways/remote_engaging_networks_test.rb
Loaded suite test/remote/gateways/remote_engaging_networks_test
Started
.........
Finished in 17.736179 seconds.
---------------------------------------------------------------------------------------------------------------------
9 tests, 18 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
---------------------------------------------------------------------------------------------------------------------
0.51 tests/s, 1.01 assertions/s

❯ docker run -it --rm -v "$PWD":/app waysact/active_merchant ruby -Itest test/unit/gateways/engaging_networks_test.rb
Loaded suite test/unit/gateways/engaging_networks_test
Started
.....
Finished in 0.0180974 seconds.
---------------------------------------------------------------------------------------------------------------------
5 tests, 19 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
---------------------------------------------------------------------------------------------------------------------
276.28 tests/s, 1049.87 assertions/s
```

henare reviewed (approved):
Really nice :trophy:
